### PR TITLE
feat: [FC-6] filter bad words

### PR DESCRIPTION
**Pull request by huydosgtech**

### CODE REVIEW

**Comment by huydosgtech** [comment]
This PR related and blocked by [FC-3](https://github.com/saigontechnology/rn-firebase-chat/pull/23).

File: src/utilities/Blacklist.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import DEFAULT_BLACKLIST from './../lang/en.json';
+
+export const badWords = (input: string) => {
```
Comment (loc-nguyenthien) [inline src/utilities/Blacklist.ts:3]:
```suggestion
export const filterBadWords = (input: string) => {
```

File: src/chat/components/InputToolbar.tsx [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            textInputStyle={[styles.textInput, composerTextInputStyle]}
+          />
+        </ScrollView>
+      </View>
+      {!!text && (
+        <PressableIcon
+          iconStyle={[styles.iconStyleDefault, iconStyle]}
+          onPress={() => {
+            if (props.isFilterBadWords) {
+              const replaceBadWords = badWords(text);
```
Comment (loc-nguyenthien) [inline src/chat/components/InputToolbar.tsx:81]:
```suggestion
              const replaceBadWords = filterBadWords(text);
```

File: src/chat/components/InputToolbar.tsx [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+  ViewStyle,
+  ImageStyle,
+} from 'react-native';
+import {
+  Composer,
+  InputToolbarProps,
+  SendProps,
+} from 'react-native-gifted-chat';
+import { PressableIcon } from './PressableIcon';
+import { badWords } from '../../utilities';
```
Comment (loc-nguyenthien) [inline src/chat/components/InputToolbar.tsx:16]:
```suggestion
import { filterBadWords } from '../../utilities';
```

File: src/utilities/Blacklist.ts (hunk 1 above)
Comment (huydosgtech) [reply src/utilities/Blacklist.ts]: ok I updated

File: src/chat/components/InputToolbar.tsx (hunk 2 above)
Comment (huydosgtech) [reply src/chat/components/InputToolbar.tsx]: ok I updated

File: src/chat/components/InputToolbar.tsx (hunk 3 above)
Comment (huydosgtech) [reply src/chat/components/InputToolbar.tsx]: ok I updated. Pls check it again

File: src/utilities/Blacklist.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import DEFAULT_BLACKLIST from './../lang/en.json';
```
Comment (baonguyenhsts) [inline src/utilities/Blacklist.ts:1]: the en.json represents localization not DEFAULT_BLACKLIST

File: src/chat/components/PressableIcon.tsx
Comment (baonguyenhsts) [inline src/chat/components/PressableIcon.tsx:1]: Why do we have to add this custom component inside a library?

**Review by baonguyenhsts — CHANGES_REQUESTED** [review]

File: src/chat/components/PressableIcon.tsx
Comment (huydosgtech) [reply src/chat/components/PressableIcon.tsx:1]: this file no need in this PR, I removed it

File: src/utilities/Blacklist.ts (hunk 4 above)
Comment (huydosgtech) [reply src/utilities/Blacklist.ts]: I changed to another name. Please help me check it

File: src/utilities/Blacklist.ts [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import badWords from '../bad_words/key.json';
+
+interface BadWords {
+  [key: string]: string[];
+}
+
+export const filterBadWords = (input: string, lang: string) => {
+  const badWordsData: string[] = (badWords as BadWords)[lang] ?? badWords.en;
+
+  const regexPattern = badWordsData
```
Comment (honghoangsts) [inline src/utilities/Blacklist.ts:10]: create this regex outside filter function

File: src/chat/components/InputToolbar.tsx [hunk 6]
```diff
@@ -83,7 +86,14 @@ const InputToolbar: React.FC<IInputToolbar> = ({
       {!!text && (
         <PressableIcon
           iconStyle={flattenedIconStyle}
-          onPress={() => onSend?.({ text: text }, true)}
+          onPress={() => {
+            if (isFilterBadWords) {
+              const replaceBadWords = filterBadWords(text, 'en');
```
Comment (honghoangsts) [inline src/chat/components/InputToolbar.tsx:91]: onSend?.({ text: filterBadWords(text, 'en') }, true);

**Review by honghoangsts — CHANGES_REQUESTED** [review]

File: src/bad_words/key.json [hunk 7]
```diff
@@ -0,0 +1,27 @@
+{
+    "en":
```
Comment (honghoangsts) [inline src/bad_words/key.json:2]: we need a function to add more bad words

**Review by honghoangsts — CHANGES_REQUESTED** [review]

File: src/bad_words/key.json (hunk 7 above)
Comment (loc-nguyenthien) [reply src/bad_words/key.json:2]: Maybe an `extraBadWords` prop on `ChatProvider` component, and move `isFilterBadWords` prop on `ChatScreen` to `ChatProvider` component too

**Review by loc-nguyenthien — CHANGES_REQUESTED** [review]

File: src/chat/components/InputToolbar.tsx (hunk 6 above)
Comment (huydosgtech) [reply src/chat/components/InputToolbar.tsx]: I removed it. It's no need

File: src/utilities/Blacklist.ts (hunk 5 above)
Comment (huydosgtech) [reply src/utilities/Blacklist.ts]: ok. I created

File: src/bad_words/key.json (hunk 7 above)
Comment (huydosgtech) [reply src/bad_words/key.json]: I tried to add isFilterBadWords in ChatProvider but it's redundant. We just need pass an array `blackListWords` and then enable filter black list

Comment (loc-nguyenthien) [reply src/bad_words/key.json]: Then we won't need the prop `isFilterBadWords` we only need to config an `blackListWords` prop on ChatProvider is enough, if we need to filter we only have to add the list there
PS. This is same with my comment below

**Review by loc-nguyenthien — APPROVED** [review]

File: src/services/firebase/firestore.ts [hunk 8]
```diff
@@ -66,10 +68,20 @@ export class FirestoreServices {
...
+    enableEncrypt,
+    encryptKey,
+    blackListWords,
+  }: FirestoreProps) => {
     if (userInfo) {
       this.userInfo = userInfo;
     }
+
+    if (blackListWords) {
+      this.blackListWords = blackListWords;
```
Comment (honghoangsts) [inline src/services/firebase/firestore.ts:82]: where did you use `this.blackListWords`

File: src/chat/ListConversationScreen.tsx [hunk 9]
```diff
@@ -22,13 +23,29 @@ export const ListConversationScreen: React.FC<IListConversationProps> = ({
...
+        if (
+          item.latestMessage?.text.length === 0 ||
+          item.latestMessage?.type !== MessageTypes.text
+        )
+          return item;
+        return {
+          ...item,
+          latestMessage: {
+            ...item.latestMessage,
+            text: filterBadWords(item.latestMessage.text, blackListWords),
```
Comment (honghoangsts) [inline src/chat/ListConversationScreen.tsx:41]: we should mask the text at formatMessage function

**Review by honghoangsts — CHANGES_REQUESTED** [review]

**Review by honghoangsts — APPROVED** [review]

**Pull request merged**